core/services/ocr2/plugins/llo/helpers_test.go

```diff
@@ -0,0 +1,370 @@
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	commonconfig "github.com/smartcontractkit/chainlink-common/pkg/config"
+
+	"github.com/smartcontractkit/chainlink/v2/core/bridges"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/cltest"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/cltest/heavyweight"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils/keystest"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/chainlink"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/chaintype"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/csakey"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/ocr2key"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/p2pkey"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/validate"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocrbootstrap"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/wsrpc/pb"
+	"github.com/smartcontractkit/chainlink/v2/core/services/streams"
+	"github.com/smartcontractkit/chainlink/v2/core/store/models"
+)
+
+var _ pb.MercuryServer = &mercuryServer{}
+
+type request struct {
+	pk  credentials.StaticSizedPublicKey
+	req *pb.TransmitRequest
+}
+
+func (r request) TransmitterID() ocr2types.Account {
+	return ocr2types.Account(fmt.Sprintf("%x", r.pk))
+}
+
+type mercuryServer struct {
+	privKey     ed25519.PrivateKey
+	reqsCh      chan request
+	t           *testing.T
+	buildReport func() []byte
+}
+
+func NewMercuryServer(t *testing.T, privKey ed25519.PrivateKey, reqsCh chan request, buildReport func() []byte) *mercuryServer {
+	return &mercuryServer{privKey, reqsCh, t, buildReport}
+}
+
+func (s *mercuryServer) Transmit(ctx context.Context, req *pb.TransmitRequest) (*pb.TransmitResponse, error) {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, errors.New("could not extract public key")
+	}
+	r := request{p.PublicKey, req}
+	s.reqsCh <- r
+
+	return &pb.TransmitResponse{
+		Code:  1,
+		Error: "",
+	}, nil
+}
+
+func (s *mercuryServer) LatestReport(ctx context.Context, lrr *pb.LatestReportRequest) (*pb.LatestReportResponse, error) {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, errors.New("could not extract public key")
+	}
+	s.t.Logf("mercury server got latest report from %x for feed id 0x%x", p.PublicKey, lrr.FeedId)
+
+	out := new(pb.LatestReportResponse)
+	out.Report = new(pb.Report)
+	out.Report.FeedId = lrr.FeedId
+
+	report := s.buildReport()
+	payload, err := mercury.PayloadTypes.Pack(evmutil.RawReportContext(ocrtypes.ReportContext{}), report, [][32]byte{}, [][32]byte{}, [32]byte{})
+	if err != nil {
+		panic(err)
```

Comment:
This one seems too severe too.

Reply:
Its in a test

Reply:
Why is it better to crash the whole test suite rather than log an error for the specific failing test?

Reply:
Yup, changed it